### PR TITLE
[hw] Add support for a Mage docker image with a Hadoop Yarn cluster

### PR DESCRIPTION
**Pull request by csharplus**
# Summary

For testing Spark sessions running on Hadoop Yarn with Mage, additional docker image setup is needed.

# Tests

Manually tested the changes locally with the following procedure:

(1) In a terminal, switch to the following folder in the Mage code repository:
```
% cd integrations/hadoop/
```
(2) Build the docker image with the following command:
```
% docker build -t mage_hadoop .
```
(3) Type this command in your terminal to start Mage using docker:
```
docker run -it --name mage_hadoop -p 6789:6789 -v $(pwd):/home/src mage_hadoop \
  /app/run_app_with_hadoop.sh mage start demo_project
```
(4) Change the `metadata.yaml` file in the main pipeline folder to include the following Spark settings:
```
spark_config:
  # Application name
  app_name: 'my spark yarn app'
  # Master URL to connect to
  # e.g., spark_master: 'spark://host:port', or spark_master: 'yarn'
  spark_master: 'yarn'
  # Executor environment variables
  # e.g., executor_env: {'PYTHONPATH': '/home/path'}
  executor_env: {}
  # Jar files to be uploaded to the cluster and added to the classpath
  # e.g., spark_jars: ['/home/path/example1.jar']
  spark_jars: []
  # Path where Spark is installed on worker nodes,
  # e.g. spark_home: '/usr/lib/spark'
  spark_home: null
  # List of key-value pairs to be set in SparkConf
  # e.g., others: {'spark.executor.memory': '4g', 'spark.executor.cores': '2'}
  others: {}
```
(5) Create a new `Standalone (batch)` pipeline and add a `Data loader`, then run it with the following code:
```
@data_loader
def load_data(*args, **kwargs):
    """
    Template code for loading data from any source.

    Returns:
        Anything (e.g. data frame, dictionary, array, int, str, etc.)
    """
    # Specify your data loading logic here    
    spark = kwargs.get('spark')
    spark_conf = spark.sparkContext.getConf()
    print('#' * 40)
    print(spark_conf.toDebugString())
    print('#' * 40)
    print(spark.sql('select 1'))
    return {}
```
(6) Verify that the added Spark code is running through a Yarn master, similar to the following: 
Screenshot 2023-05-24 at 1 15 41 AM

The steps to set up local Hadoop and Spark on Mac are listed at [Hadoop + Spark installation (OSX)](https://gist.github.com/cjzamora/9fcc740228875f4642a1), and summarized below (it is not used for testing this PR,  though it could be potentially useful in the future):
1. Install Homebrew following the instructions given at [Homebrew](https://brew.sh/), or run the following command in a terminal:
```
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
2. Install Java:
```
% brew install openjdk@11
```
3. Install Hadoop (with Yarn)
```
% brew install hadoop
```
4. Install Apache Spark (and dependencies):
```
# Install Apache Spark
% brew install apache-spark scala sbt
```
5. Open `~/.zshrc` and add the following settings (replace the Hadoop version `3.3.4` to the right one when needed):
```
# Added for Hadoop
export HADOOP_VERSION=3.3.4
export HADOOP_HOME=/usr/local/Cellar/hadoop/3.3.4/libexec
export HADOOP_CONF_DIR=$HADOOP_HOME/etc/hadoop/
export PATH=$HADOOP_HOME/bin:$HADOOP_HOME/sbin:$PATH

alias hstart="/usr/local/Cellar/hadoop/3.3.4/sbin/start-all.sh"
alias hstop="/usr/local/Cellar/hadoop/3.3.4/sbin/stop-all.sh"
```
Run the following command to set up the required environment variables:
```
% source ~/.zshrc
```
6. For Hadoop to run on Mac OSX, Remote Login and `File Sharing` under `System Preferences` also need to be turned on. Configure `passphraseless` SSH on localhost if it is not already set up, using the following steps:
```
i). ssh-keygen -t rsa
Press enter for each line 
ii). cat ~/.ssh/id_rsa.pub >> ~/.ssh/authorized_keys
iii). chmod og-wx ~/.ssh/authorized_keys
```
7. Start the Hadoop and Yarn cluster with the following command:
```
% hstart
```
9. Run `spark-shell` from the command line to launch Spark shell for Scala for testing:
```
% spark-shell
```
. Create a Spark `DataFrame` with some sample data to validate the installation:
```
import spark.implicits._
val data = Seq(("Java", "20000"), ("Python", "100000"), ("Scala", "3000"))
val df = data.toDF() 
df.show()
```
10. To start the Spark shell for Python, run the following command:
```
% ./bin/pyspark
```
Or if `PySpark` is installed with pip in your current environment:
```
% pyspark
```
11. Create a Python test script `pi.py` with the following code:
```
import sys
from random import random
from operator import add

from pyspark.sql import SparkSession


if __name__ == "__main__":
    """
        Usage: pi [partitions]
    """
    spark = SparkSession\
        .builder\
        .appName("PythonPi")\
        .getOrCreate()

    partitions = int(sys.argv[1]) if len(sys.argv) > 1 else 2
    n = 100000 * partitions

    def f(_: int) -> float:
        x = random() * 2 - 1
        y = random() * 2 - 1
        return 1 if x ** 2 + y ** 2 <= 1 else 0

    count = spark.sparkContext.parallelize(range(1, n + 1), partitions).map(f).reduce(add)
    print("Pi is roughly %f" % (4.0 * count / n))

    spark.stop()
```
12. Run the following command to test `spark-submit` with `pi.py`:
```
% spark-submit pi.py
```
13. To run `pi.py` with the local Hadoop Yarn cluster, run the following command:
```
% spark-submit --master yarn --name test_job pi.py
```
It should generate outputs similar to the following
```
23/05/20 17:15:22 INFO DAGScheduler: Job 0 finished: reduce at /Users/hw/workspace/spark/pi.py:42, took 2.374661 s
Pi is roughly 3.142480
```

cc:

@dy46 @wangxiaoyou1993

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
How do we add the custom SparkConf in project's metadata.yaml? Could you provide an example?

**Comment by csharplus** [comment]
> How do we add the custom SparkConf in project's metadata.yaml? Could you provide an example?

I've add a setting section to two template files. Please see if these two templates are the right place to add these settings.
(1) [mage_ai/data_preparation/templates/main/metadata.yaml](https://github.com/mage-ai/mage-ai/pull/2664/files#diff-64069f06e597393432882c498fd550827ef9f873aea406439fa8d55d230f30cc)
(2) [mage_ai/data_preparation/templates/repo/metadata.yaml](https://github.com/mage-ai/mage-ai/pull/2664/files#diff-c952d622752b328854a4f5be3a876be493b2cea3b67a19e62e4e4a038882d9f0)
```
spark_config:
  # Application name
  app_name: 'my spark app'
  # Master URL to connect to
  # e.g., spark_master: 'spark://host:port', or spark_master: 'yarn'
  spark_master: 'local'
  # Jar files to be uploaded to the cluster and added to the classpath
  # e.g., spark_jars: ['/home/path/example1.jar']
  spark_jars: []
  # Path where Spark is installed on worker nodes,
  # e.g. spark_home: '/usr/lib/spark'
  spark_home: null
  # List of key-value pairs to be set in SparkConf
  # e.g., others: {'spark.executor.memory': '4g', 'spark.executor.cores': '2'}
  others: {}
```
For SparkSession, it was designed to be a singleton instance re-used for the whole application session somehow, which made it difficult to switch to another session with a different set of parameters.

File: mage_ai/data_preparation/models/block/__init__.py [hunk 1]
```diff
@@ -1784,11 +1786,33 @@ def __get_spark_session(self):
...
+            spark_config = repo_config.get('spark_config')
+
+            from pyspark.conf import SparkConf
             from pyspark.sql import SparkSession
-            self.spark = SparkSession.builder.master(
-                os.getenv('SPARK_MASTER_HOST', 'local')).getOrCreate()
+
+            if spark_config:
+                conf = SparkConf()
+                if repo_config.get('app_name'):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/block/__init__.py:1797]: Did you test this code? It should be `spark_config.get('xxx')`, right?

Comment (csharplus) [reply mage_ai/data_preparation/models/block/__init__.py:1797]: Good catch: made this fix. I encountered some difficulty with connecting to a Yarn cluster to test out this code path so far: (i) from docker to my local Yarn cluster; (ii) setting up a Hadoop/Yarn together with Mage.

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**